### PR TITLE
Extends SyncMap with new utility methods

### PR DESCRIPTION
**Pull request by tcsc**
Adds some useful utility methods to the the SyncMap class.

The motivating example for these additional methods is the Okta AccessList reconciler,
which is currently using several map+mutex pairs and lots of boilerplate, which has been factored out into these new methods.

### CODE REVIEW

**Review by zmb3 — CHANGES_REQUESTED** [review]
We already have a sync.Map implementation

**Comment by tcsc** [comment]
From the [`sync.Map` documentation](https://pkg.go.dev/sync#Map)

> The Map type is specialized. Most code should use a plain Go map instead, with separate locking or coordination, for better type safety and to make it easier to maintain other invariants along with the map content.

The intent here is to fill that type safety niche in places where we are currently using multiple pairs of maps & mutexes. I'm also reasonably sure that the intended use cases don't really fall in to either of the cases that `sync.Map` is optimized for.

I'm happy to defer to an existing typesafe `sync.Map` implementation, but didn't see any obvious contenders in our codebase (which, of course, doesn't mean that they're not there...)

I can also add some documentation on the differences between this and `sync.Map` if that will help.

File: lib/utils/sync_map.go [hunk 1]
```diff
@@ -78,16 +78,75 @@ func (s *SyncMap[K, V]) LoadAndDelete(key K) (value V, ok bool) {
...
+	defer s.mu.Unlock()
+
+	if s.values != nil {
+		clear(s.values)
+	}
+}
+
+// Set takes ownership of a map, replacing the existing managed map with the
+// supplied value
+func (s *SyncMap[K, V]) Set(m map[K]V) {
```
Comment (zmb3) [inline lib/utils/sync_map.go:107]: I wonder if this use case would be better served by a constructor.

Also, I understand that your inspiration for some of these utilities is Rust, but the reality is Go does not have Rust's ownership semantics and there is no such thing as "taking ownership of a map" in Go. The caller of this method has just as much ownership over `m` before and after this method runs.

**Review by zmb3 — COMMENTED** [review]
No major concerns, but I don't love this either. I think you're almost always better off managing your own map with the locking patterns that you need for your particular use case than trying to build a generic one-size-fits-all solution.

File: lib/utils/sync_map.go (hunk 1 above)
Comment (tcsc) [reply lib/utils/sync_map.go:107]: I see where you're coming from, but there are cases in Okta Reconciler clients that currently wants to entirely replace the map under management post-construction. 

And perhaps "ownership" is not the right word - it certainly has overtones from Rust that are not directly applicable to Go. I'm just trying to convey the idea that all bets are off if you directly modify `m` outside after calling this. Maybe just  spelling it out is better.

**Comment by tcsc** [comment]
>  I think you're almost always better off managing your own map with the locking patterns that you need for your particular use case

I can understand that. Let me go back a step, and go back to the motivation for doing this in the first place.  The specific set of cases I'm trying to tidy up is our reconciler code. Each reconciler uses at least two map+mutex pairs, and in some cases there are several reconcilers working together, which also multiplies the number of pairs required.

I find most code using the Reconcilers incredibly hard to follow thanks to the sheer amount of lock/unlock boilerplate. In the more complex cases we're just _asking_ to lock the wrong map with the wrong mutex. With so many map+mutex pairs in flight I have no confidence that review will catch a mismatch.

I'm sympathetic to the argument that "if things are difficult because they're complicated then we should split them up so they're less complicated", but that requires specialised knowledge of each use reconciler that I don't have, plus probably more engineering time than I have available. I think that tying the map+lock pairs together _explicitly_ (as opposed to being tied together _implicitly_, as they are now) makes things less bad, and this PR is part of doing that.

I also get that this motivating example (i.e.refactoring the reconcilers) is a specific use case - but unfortunately one that is spread around Teleport anywhere that uses the `services.Reconciler`. I'm happy to consider creating a locked map type specifically for use with the reconciler, but then we have two `SyncMap` types. 

I've already tried using simple functions to do this (e.g. `func lockedMapClone(*sync.Mutex, map[K]V){...}`, and while it solves some of the readability and boilerplate issues, it still requires a human to make sure that its invoked with the right map+mutex pair.

So that's what I'm trying to do in detail, and I'm happy to consider any advice or guidance.

**Review by zmb3 — APPROVED** [review]

**Review by r0mant — APPROVED** [review]
